src/backend/x11/window.rs

```diff
@@ -0,0 +1,259 @@
+    pub fn gc(&self) -> u32 {
+        if let Some(inner) = self.0.upgrade() {
+            inner.gc
+        } else {
+            0
+        }
+    }
+}
+
+impl From<Arc<WindowInner>> for Window {
+    fn from(inner: Arc<WindowInner>) -> Self {
+        Window(Arc::downgrade(&inner))
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct WindowInner {
+    pub connection: Arc<XConnection>,
+    pub inner: x11::Window,
+    root: x11::Window,
+    pub atoms: Atoms,
+    pub size: Mutex<Size<u16, Logical>>,
+    pub depth: Depth,
+    pub gc: x11::Gcontext,
+}
+
+impl WindowInner {
+    pub fn new(
+        connection: Arc<XConnection>,
+        screen: &Screen,
+        properties: WindowProperties<'_>,
+        atoms: Atoms,
+        depth: Depth,
+        visual_id: u32,
+        colormap: u32,
+    ) -> Result<WindowInner, X11Error> {
+        let xcb = connection.xcb_connection();
+
+        // Generate the xid for the window
+        let window = xcb.generate_id()?;
+        let window_aux = CreateWindowAux::new()
+            .event_mask(
+                EventMask::EXPOSURE // Be told when the window is exposed
+            | EventMask::STRUCTURE_NOTIFY
+            | EventMask::KEY_PRESS // Key press and release
+            | EventMask::KEY_RELEASE
+            | EventMask::BUTTON_PRESS // Mouse button press and release
+            | EventMask::BUTTON_RELEASE
+            | EventMask::POINTER_MOTION // Mouse movement
+            | EventMask::RESIZE_REDIRECT // Handling resizes
+            | EventMask::NO_EVENT,
+            )
+            // Border pixel and color map need to be set if our depth may differ from the root depth.
+            .border_pixel(0)
+            .colormap(colormap);
+
+        let cookie = xcb.create_window(
+            depth.depth,
+            window,
+            screen.root,
+            0,
+            0,
+            properties.width,
+            properties.height,
+            0,
+            WindowClass::INPUT_OUTPUT,
+            visual_id,
+            &window_aux,
+        )?;
+
+        // Send requests to change window properties while we wait for the window creation request to complete.
+        let mut window = WindowInner {
+            connection: connection.clone(),
+            inner: window,
+            root: screen.root,
+            atoms,
+            size: Mutex::new((properties.width, properties.height).into()),
+            depth,
+            gc: 0,
+        };
+
+        let gc = xcb.generate_id()?;
+        xcb.create_gc(gc, window.inner, &CreateGCAux::new())?;
+        window.gc = gc;
+
+        // Enable WM_DELETE_WINDOW so our client is not disconnected upon our toplevel window being destroyed.
+        xcb.change_property32(
+            PropMode::REPLACE,
+            window.inner,
+            atoms.WM_PROTOCOLS,
+            AtomEnum::ATOM,
+            &[atoms.WM_DELETE_WINDOW],
+        )?;
+
+        // Block until window creation is complete.
+        cookie.check()?;
```

Comment:
...Why? I know that everyone does this, but I never understood wait. Sure, this is not really expensive, but it is still one more round-trip to the X11 server.

And if this is about the error checking: Hm. How likely is it that creating the window fails? And the error is not silently lost anyway, it would then show up on the event loop instead...